sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java

```diff
@@ -661,21 +671,22 @@ public final int appendArray(int length) {
    * common non-struct case.
    */
   public final int appendStruct(boolean isNull) {
+    reserve(elementsAppended + 1);
     if (isNull) {
       // This is the same as appendNull but without the assertion for struct types
-      reserve(elementsAppended + 1);
       putNull(elementsAppended);
-      elementsAppended++;
       for (WritableColumnVector c: childColumns) {
-        if (c.type instanceof StructType) {
+        if (c.isStruct()) {
           c.appendStruct(true);
         } else {
           c.appendNull();
         }
       }
     } else {
-      appendNotNull();
+      putNotNull(elementsAppended);
     }
+    putStruct(elementsAppended, elementsAppended);
```

Comment:
For `isNull` case, we also need do `putStruct`?

Reply:
I think the doc of `putStruct` is `Puts a new non-null struct ...`

Reply:
For the `isNull` case, clients should just call `putNull` or `putNulls`. I think this is the same as other types like int, boolean, etc.

Reply:
Oops sorry I misunderstood you. Yes I think we can move this to the not-null branch. for `isNull` case it is unnecessary.